src/Microsoft.AspNetCore.Routing/Matching/EndpointComparer.cs

```diff
@@ -0,0 +1,162 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Routing.Matching
+{
+    // Use to sort and group Endpoints. RouteEndpoints are sorted before other implementations.
+    //
+    // NOTE:
+    // When ordering endpoints, we compare the route templates as an absolute last resort.
+    // This is used as a factor to ensure that we always have a predictable ordering
+    // for tests, errors, etc.
+    //
+    // When we group endpoints we don't consider the route template, because we're trying
+    // to group endpoints not separate them.
+    //
+    // TLDR: 
+    //  IComparer implementation considers the template string as a tiebreaker.
+    //  IEqualityComparer implementation does not.
+    //  This is cool and good.
+    internal class EndpointComparer : IComparer<Endpoint>, IEqualityComparer<Endpoint>
+    {
+        private readonly IComparer<Endpoint>[] _comparers;
+
+        public EndpointComparer(IEndpointComparerPolicy[] policies)
+        {
+            // Order, Precedence, (others)...
+            _comparers = new IComparer<Endpoint>[2 + policies.Length];
+            _comparers[0] = OrderComparer.Instance;
+            _comparers[1] = PrecedenceComparer.Instance;
+            for (var i = 0; i < policies.Length; i++)
+            {
+                _comparers[i + 2] = policies[i].Comparer;
+            }
+        }
+
+        public int Compare(Endpoint x, Endpoint y)
+        {
+            // We don't expose this publicly, and we should never call it on
+            // a null endpoint.
+            Debug.Assert(x != null);
+            Debug.Assert(y != null);
+
+            var compare = CompareCore(x, y);
+
+            // Since we're sorting, use the route template as a last resort.
+            return compare == 0 ? ComparePattern(x, y) : compare;
+        }
+
+        private int ComparePattern(Endpoint x, Endpoint y)
+        {
+            var routeEndpointX = x as RouteEndpoint;
+            var routeEndpointY = y as RouteEndpoint;
+
+            if (routeEndpointX != null)
+            {
+                if (routeEndpointY != null)
+                {
+                    return routeEndpointX.RoutePattern.RawText.CompareTo(routeEndpointY.RoutePattern.RawText);
+                }
+
+                return 1;
+            }
+            else if (routeEndpointY != null)
+            {
+                return -1;
+            }
+
+            return 0;
```

Comment:
This check that the Endpoint is a RouteEndpoint is a bit verbose in each compare method :(

Reply:
yeah :-/

Reply:
Hmmm, I wonder if these comparers should return 0 if both endpoints are not RouteEndpoint, and then this logic is handled in the final pattern comparison. That will let `IEndpointComparerPolicy` have a chance to have logic for comparing a RouteEndpoint with a non-RouteEndpoint.

What do you think?

Reply:
>  have a chance to have logic for comparing a RouteEndpoint with a non-RouteEndpoint.

I'm not sure how useful it is. The use case for non-`RouteEndpoint` is to inject it as a failure case. It seems natural to me to sort those to the end of the list. 

I would expect that a `RouteEndpoint` > (other) for all cases